dma: stm32: driver refactoring

lowlander opened the pull request:
DMA driver cleanup

- Remove the use of k_malloc 
- Change way how ISR gets the channel ID
- Remove global (non-namespaced) channel/stream lookup tables
- Remove global (non-namespaced) function tables
- Small code cleanups
- Small updates to tests/drivers/dma/loop_transfer

Tested with SPI FRAM on nucleo-L496ZG, nucleo-F767ZI, and nucleo-L053r8
Tested with tests/drivers/dma/loop_transfer on nucleo-F207zg, nucleo-L496ZG, nucleo-F767ZI, and nucleo-L053r8
Tested with tests/drivers/spi/spi_loopback (in combination with PR #27263) on nuclueo-L496ZG, nucleo-F767ZI, and nucleo-L053r8
Tested with OLED on custom STM32F407 board

lowlander committed:
drivers: dma: stm32: refactor irq handling

- Remove the need for kmalloc
- On hardware that supports it use 1 IRQ handler per stream to
  determine the stream ID, so the ISR does not have to loop
  over all ID's to see which one is active. On hardware (like
  STM32L0 and STM32F0) where up to 7 streams share 3 IRQ's use
  the loop to check which stream is active.

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lowlander committed:
drivers: dma: stm32: convert global conversion tables to functions

Convert global conversion tables for id-to-stream and slot-to-channel
to functions that use local const static conversion tables.

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lowlander committed:
drivers: dma: stm32: Use volatile for flag used in ISR

Use volatile for flag shared between normal code and ISR to
prevent the compiler from possibly optimizing it away.

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lowlander committed:
drivers: dma: stm32: Improve code readability

Some small code readability changes.

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lowlander commented:
@nandojve could you give this a try? PS: do you have an example app that uses SPI/DMA, in the normal zephyr samples I could not find anything that really worked.

gmarull commented on drivers/dma/dma_stm32.h:
```diff
@@ -7,6 +7,30 @@
 #ifndef DMA_STM32_H_
 #define DMA_STM32_H_
 
+#if DT_NODE_HAS_STATUS(DT_DRV_INST(0), okay)
```
headers should ideally be self-contained, so devicetree API should be included here

lowlander replied:
Probably could be move to the .c file.

gmarull commented on drivers/dma/dma_stm32.c:
```diff
@@ -227,10 +223,9 @@ static int dma_stm32_get_periph_increment(enum dma_addr_adj increment,
-	struct dma_stm32_stream *stream = &data->streams[id - STREAM_OFFSET];
-	const struct dma_stm32_config *dev_config =
-					dev->config;
+	const struct dma_stm32_config *dev_config = dev->config;
```
`config` name would be better to keep code namings consistent.

lowlander replied:
There is/was already a `config` function argument, and I didn't want to change that.

gmarull replied:
Sorry, missed that!

gmarull commented on drivers/dma/dma_stm32.c:
```diff
@@ -54,26 +54,15 @@ static void dma_stm32_clear_stream_irq(struct device *dev, uint32_t id)
-
-	if (id == data->max_streams) {
-		LOG_ERR("Unknown interrupt happened.");
+	if (id >= config->max_streams) {
```
Maybe an assertion would be better? Is there any chance for this to happen?

lowlander replied:
It should not happen AFAIK, so maybe a ASSERT is OK.

gmarull reviewed (commented):
LGTM, added some comments

lowlander commented:
Note: next week I should have L0 and L4 hardware so I can test both V1 and V2 hardware, which should make development easier and a bit faster.

gmarull commented:
@lowlander I have H7, F4, F3 and G4 for testing if needed.

lowlander commented:
> @lowlander I have H7, F4, F3 and G4 for testing if needed.

@gmarull F407 is my main target, H7, F3, and G4 all have the same setup (one IRQ per stream) as the F4, right ? But testing is welcome, especially with different hardware than SPI (the only driver I have the uses DMA)

lowlander commented:
@gmarull I updated the code as you requested, but until I have L0 hardware this is still DNM/WIP

lowlander commented:
@gmarull I got things to work on a L053R8 board with a SPI-FRAM, maybe you can give this (and if you do SPI #27263) a try and see if you can break it.

lowlander commented on drivers/dma/dma_stm32.c:
```diff
@@ -13,17 +13,46 @@
+    defined(CONFIG_SOC_SERIES_STM32F0X)
+#define DMA_STM32_SHARED_IRQS
+#endif
+
```
@gmarull @FRASTM this (the checking if IRQ's are shared) I  dislike because, would it be better to do that in Kconfig ? Or do you have any better ideas, on how to deal with the shared irq's at compile time ?

cybertale replied:
Hi, I think if this is a IP block specific feature, we can use IP block specific kconfig symbols to do the job. We can define them in soc/arm/st_stm32/stm32xx/Kconfig.defconfig.series like the way DMA_STM32_v1 and DMA_STM32_V2 are defined. For example we define DMA_STM32_SHARED_IRQ in some socs and not in others. So in the driver context, we can do:
```
#if defined(CONFIG_DMA_STM32_SHARED_IRQ)
...
#else
...
#endif
```
But this should depend on if we do have too many socs that just using several kinds of IP blocks.

lowlander replied:
Yes I thought about that, but didn't want to touch all SOC Kconfig files, before I knew things actually worked.  I also tried to think of some "clever" way to get the info out of the DT, but doing that at compile time with the C-preprocessor is not that easy (I might give it a second try, because the info we need is in DT).

cybertale replied:
I just remember this info is retrieved from DT. You mean the python script? Using a function to determine if there are multiple lines use the same interrupt line, and create a new symbol in generated DT files to indicate if this one has shared IRQs? I think this surely is the clever way!

lowlander replied:
Maybe in the future we can add a maybe-shared-flag to the `IRQ_CONNECT` macro, or make a `IRQ_CONNECT_SHARED` macro. But that is outside the scope of this PR.

For the moment I first solved it by putting it in Kconfig,  this PR is already big enough :-)

aurel32 commented:
I tested this on a L433 with the SX1276 lora driver in combination with #27263 and it works all fine. Thanks.

FRASTM commented:
there is a test in the tests/drivers/spi/spi_loopback to transfer on the SPI bus, when the Tx pin is connected to the Rx on the nucleo board.

lowlander commented:
> there is a test in the tests/drivers/spi/spi_loopback to transfer on the SPI bus, when the Tx pin is connected to the Rx on the nucleo board.

will try it out and report back

lowlander committed:
drivers: dma: stm32: prevent posible out of bound access

Set pointer to an array element only after checking the bounds
of the index.

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lowlander committed:
drivers: dma: stm32: check result of stream disabling

Use dma_stm32_disable_stream instead of stm32_dma_disable_stream
to check if the disabling of the stream really worked and if
not return an error.

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lowlander committed:
drivers: dma: stm32: remove global function tables

Convert the driver from global function tables to functions
that use static const function tables.

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

nandojve commented:
@lowlander good news!

It worked with S2LP and M95640, max speed 7.5MHz
I'll do configs to test STM32L072CZ, STM32L475VG and STM32F767ZI too. I'll run this only tomorrow.

I think we will need send DMA configs and enable tests for build all configs.
How do you are planning this?

Anyway, I hope we can get this merged soon.

Thank you so much!

nandojve commented:
@lowlander, I already tested all boards. L053, L072, L475 are OK.

Then, the try with F767 with below config not worked. I'm not sure it is right.
```
	dmas = <&dma2 5 3 0x28440 0x03>,
	       <&dma2 2 3 0x28480 0x03>;
```

I got this when enable logs
```
D: Installed config 0x20020160: freq 562500Hz (div = 128), mode 0/0/0, slave 1
D: tx_bufs 0x20021710 - rx_bufs 0x20021720 - 1
D: current_tx 0x20021708 (1), current_rx 0x20021718 (1), tx buf/len 0x20021700/2, rx buf/len 0x20021704/2
D: tx buf/len 0x00000000/0
D: rx buf/len 0x00000000/0
D: tx_bufs 0x20021700 - rx_bufs 0x20021708 - 1
D: current_tx 0x200216f8 (1), current_rx 0x20021710 (2), tx buf/len 0x200216f0/2, rx buf/len 0x200216f4/2
D: tx buf/len 0x00000000/0
D: rx buf/len 0x20021734/2
D: rx buf/len 0x00000000/0
D: tx_bufs 0x20021728 - rx_bufs 0x20021730 - 1
D: current_tx 0x20021720 (1), current_rx 0x20021738 (2), tx buf/len 0x20021718/2, rx buf/len 0x2002171c/2
D: tx buf/len 0x00000000/0
D: rx buf/len 0x20021717/1
D: rx buf/len 0x00000000/0
D: tx_bufs 0x20021728 - rx_bufs 0x20021730 - 1
D: current_tx 0x20021720 (1), current_rx 0x20021738 (2), tx buf/len 0x20021718/2, rx buf/len 0x2002171c/2
D: tx buf/len 0x00000000/0
D: rx buf/len 0x20021717/1
D: rx buf/len 0x00000000/0
```

lowlander commented:
> @lowlander, I already tested all boards. L053, L072, L475 are OK.
> 
> Then, the try with F767 with below config not worked. I'm not sure it is right.
> 

@nandojve the F767 is a M7 and there seem to be two big differences with the others
1) it has a data cache, that might interfere with DMA
2) the DMA controller seems a bit different (it has a double buffer mode)

I have a F767 nucleo and I see "valid" SPI on the bus, just the data is wrong. I will have to investigate what the real differences are and how to deal with the cache.

nandojve commented:
@lowlander yes, you are right!

Zephyr allows you to allocate memory that will not be part of data cache. I remember that I need handler that thing when port SAM4E Ethernet support just because cache were enabled for SAME70/V71.
I think the magic starts here
`/arch/arm/core/aarch32/Kconfig:16:	select ARCH_HAS_NOCACHE_MEMORY_SUPPORT if ARM_MPU && CPU_HAS_ARM_MPU && CPU_CORTEX_M7`

You can get inspiration looking at https://github.com/zephyrproject-rtos/zephyr/blob/9eb9c4cce9b04f8fbe930f93bc97638fe31d3e3b/drivers/ethernet/Kconfig.sam_gmac#L9 This will enable NOCACHE sections on linker. Then, you need only define the memory that will not be part of the dcache https://github.com/zephyrproject-rtos/zephyr/blob/9eb9c4cce9b04f8fbe930f93bc97638fe31d3e3b/drivers/ethernet/eth_sam_gmac.c#L130-L131

You can enable defines that allows you to invalidate cache when necessary, Start look from here: https://github.com/zephyrproject-rtos/zephyr/blob/9eb9c4cce9b04f8fbe930f93bc97638fe31d3e3b/drivers/ethernet/eth_sam_gmac.c#L55

At DMA loop test, there are implicit sections to define how avoid dcache. This is another way to define I think: https://github.com/zephyrproject-rtos/zephyr/blob/9eb9c4cce9b04f8fbe930f93bc97638fe31d3e3b/tests/drivers/dma/loop_transfer/src/dma.c#L22-L33

Well, I think SAM eth driver is a good start point. I hope this could bring some ideas.

lowlander commented:
@nandojve I updated both PR's, now F7 should also work when using cacheable memory, if the new CONFIG_SPI_STM32_UNALIGNED_DMA=y option is set. That causes the SPI driver to split up buffers to make sure they are aligned with the cache-line. If that is not possible it copies the data to a temp buffer that is aligned. Could you give both PR's a try again on your different hardwares?

nandojve commented:
> @nandojve I updated both PR's, now F7 should also work when using cacheable memory, if the new CONFIG_SPI_STM32_UNALIGNED_DMA=y option is set. That causes the SPI driver to split up buffers to make sure they are aligned with the cache-line. If that is not possible it copies the data to a temp buffer that is aligned. Could you give both PR's a try again on your different hardwares?

Hi @lowlander, all boards are working now. Just to inform that L05/7 max speed will be less than 8MHz. Important mention that máx speed is for both cases (INT and DMA), so, I think DMA is fine and probably is a HW limitation. With L475 I got 20MHz on EEPROM and 10MHz on S2LP, prescalers match max speed with both devices. The F767 prescaler match 16 and 8MHz. I tested with LOGs too, so, it seems not be a problem anymore.

Thank you for your effort!

lowlander committed:
tests: dma: loop_transfer: allow configuration of device and channel

Add Kconfig with the following options:

- DMA_LOOP_TRANSFER_DRV_NAME to set the DMA device name, because
  some SOCs do not allow memory to memory DMA on all DMA devices
- DMA_LOOP_TRANSFER_CHANNEL_NR some SOCs start counting the DMA
  channels at 1 instead of 0.

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lowlander committed:
tests: dma: loop_transfer: add board configurations

Add configurations for Nucleo F207ZG, F767ZI, L053R8, and L496ZG.

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

lowlander commented:
@gmarull @erwango @aurel32 @FRASTM I think this PR now is in a state that I would like to ask for a review and possible merge if everything is OK.

cybertale commented on drivers/dma/dma_stm32.c:
```diff
@@ -232,13 +224,8 @@ static int dma_stm32_get_periph_increment(enum dma_addr_adj increment,
-static int dma_stm32_configure(struct device *dev, uint32_t id,
-			       struct dma_config *config)
-#endif /* CONFIG_DMAMUX_STM32 */
+DMA_STM32_EXPORT_DMAMUX int dma_stm32_configure(struct device *dev, uint32_t id,
```
I think there should be a more generic symbol here, like `DMA_STM32_EXPORT`. When I saw this symbol first time, I thought it's only used for the ones with DMAMUX features. What do you think?

lowlander replied:
Well the name was based on the original `#ifdef CONFIG_DMAMUX_STM32` . But now I even wonder if it even makes sense to add all that code just to add/remove `static` to the function. I don't think it would be a problem to just always leave away the `static`?

cybertale replied:
At least for me it's a coding habit that trying to add `static` to all the functions, making symbols less exposed in the kernel. Yes I also think this is unnecessary since we've already named the function with module name as prefix. While this is a coding rule as I remember in the Linux kernel. Zephyr is Linux's little brother, and it also inherits many Linux's coding style.

lowlander replied:
Yeah static makes sense, no question about it, but these functions are kind of a public API so I think we can remove the DMASTM32_EXPORT_DMAMUX and just always "export" them, makes the code clearer I think.

lowlander replied:
For the moment I just removed the `static` , but I think if we refactor the dmamux driver we probably can use the dma driver struct to get the functions. But I think that should be done in a different PR.

lowlander replied:
@cybertale Can we put this on "resolved" too ? Maybe relook at it when refactoring the dmamux driver ?

cybertale replied:
Of course this way it will work, but I don't know if this is acceptable. This idea of remove some existed static... I think we need some upstream maintainer's opinion like @erwango 's about this. 
This is not a big problem, should be resolved soon. Sorry I don't have much free time in work days, havn't look all your commits, but I 'll try these two days. Changes I've seen so far so good!

lowlander replied:
I have no problem putting the defines back in, maybe with a different name like maybe DMA_STM32_EXPORT_API  ?

lowlander replied:
@cybertale is this an acceptable solution? I think the `DMA_STM32_EXPORT_API` name should be clear enough for everybody and the code is more readable than with the whole ``#ifdef #else #endif``

lowlander replied:
@cybertale and @erwango is the use of `DMA_STM32_EXPORT_API` OK, or should that be done in some other way?

cybertale replied:
Yes, this is acceptable for me.

cybertale commented on drivers/dma/dma_stm32.c:
```diff
@@ -570,25 +593,16 @@ static int dma_stm32_init(struct device *dev)
 
-	int size_stream =
-		sizeof(struct dma_stm32_stream) * data->max_streams;
-	data->streams = k_malloc(size_stream);
```
Glad we removed this. One headache problem this cause is that default HEAP_MEM_POOL_SIZE in zephyr is 0, we have to manually set one, which is 1024 in this driver. But this is not the right way of doing this, we should ask for some bytes to be added to the HEAP_MEM_POOL_SIZE, which is a way Zephyr not supported(don't know if this is supported yet).
Anyway, since we removed k_malloc, we should remove size definition of HEAP_MEM_POOL_SIZE on soc/arm/st_stm32/common/Kconfig.defconfig.series:78. But I see the dma_mux driver is still using the k_malloc, we can define HEAP_MEM_POOL_SIZE when DMAMUX_STM32 is defined, which is defined on drivers/dma/Kconfig.stm32:30

cybertale replied:
Here I wasn't asking for refactoring dmamux. I also support to make it in seperate PRs. What I want to say is for the new driver, we do not need HEAP_MEM_POOL_SIZE at all, but we are still setting it to 1024. My suggestion is to keep the action of setting HEAP_MEM_POOL_SIZE to dmamux driver itself.

lowlander replied:
Where are we setting HEAP_MEM_POOL_SIZE  ?

cybertale replied:
I'm thinking about put it in drivers/dma/Kconfig.stm32:28, which means that HEAP_MEM_POOL_SIZE will be set to 1024 when the DMAMUX driver is selected.

lowlander replied:
OK but there is nothing in this PR (and the DMA driver) that selects HEAP_MEM_POOL_SIZE, is there ?

cybertale replied:
You didn't get it...   Maybe I'm not explaining this clear enough... :sweat_smile:
k_malloc in zephyr needs user to set HEAP_MEM_POOL_SIZE to a non-zero value so we can k_malloc some memory. Since this driver was using k_malloc and the default value of HEAP_MEM_POOL_SIZE in Zephyr is zero, we set HEAP_MEM_POOL_SIZE to 1024 as long as `DMA` symbol is set in ST series socs, which is done on soc/arm/st_stm32/common/Kconfig.defconfig.series:78.
But our new driver gets rid of k_malloc, thus we can get rid of action of setting HEAP_MEM_POOL_SIZE.
The ideal way would be remove it, but dmamux driver is still using k_malloc, so HEAP_MEM_POOL_SIZE have to be set when dmamux is used but leaves untouched in this driver.
This PR isn't setting HEAP_MEM_POOL_SIZE, but the original one is. This PR should let HEAP_MEM_POOL_SIZE is set to 1024 only when dmamux driver is in use.

lowlander replied:
Ah OK now I get you. I moved the mempool selection under dmamux, this should reduce memory use on systems without dmamux which is good news for @nandojve I think.

PS: I just ordered a L4R5 based board to do the DMAMUX development.

lowlander committed:
drivers: dma: stm32: only set mem pool size when using DMAMUX

The general DMA driver doesn't use kmalloc anymore so it doesn't
need a memory pool. The DMAMUX_SMT32 driver still uses kmalloc,
so move the HEAP_MEM_POOL_SIZE config under DMAMUX_STM32.

Signed-off-by: Erwin Rol <erwin@erwinrol.com>

erwango commented on drivers/dma/dma_stm32.h:
```diff
@@ -7,6 +7,11 @@
 #ifndef DMA_STM32_H_
 #define DMA_STM32_H_
 
+#include <soc.h>
+#include <drivers/dma.h>
+#include <drivers/clock_control.h>
+#include <drivers/clock_control/stm32_clock_control.h>
```
Do we need all these ?
At least in this commit, this doesn't seem obvious.

lowlander replied:
"include and forget", so probably not, the thing that was needed were the include(s) for `struct stm32_pclken` because that is used in the header file, and the idea was that zephyr includes must be self contained, right ?

Let me figure out which one is really needed and which ones aren't.

lowlander replied:
I cleaned up the #includes, hope it is OK now?

erwango reviewed (commented):
Looks nice in general, one thing to be clarified

lowlander commented:
Before this is merged (hopefully soon) and closed, thanks to all reviewers and testers, especially @nandojve

lowlander commented:
@nashif @cybertale @gmarull @FRASTM could you also "approve" this PR, so that it is clear you're all OK with it.

erwango commented:
@lowlander please rebase before going further

lowlander commented:
Nice! All out standing PR's broken beyond a simple rebase. That all breaking update couldn't have waited until PR's that were almost ready were merged, could it? Zephyr developers are not making friends like this, I am wasting lot of time this way.

cybertale reviewed (commented):
Sorry need to bother you again. There may be some typos in the commit message of commit "drivers: dma: stm32: refactor irq handling". I think the first sentence should be "On **hardwares** that **support** it use 1 IRQ", "wich one" should be "which one", and "upto" should be "up to"?
I'm also not a native English speaker, forgive me if I were wrong, just want to make it better. :-)

cybertale commented on drivers/dma/dma_stm32.c:
```diff
@@ -271,7 +271,9 @@ DMA_STM32_EXPORT_API int dma_stm32_configure(const struct device *dev,
 	}
 
-	stm32_dma_disable_stream(dma, id);
+	while (stm32_dma_disable_stream(dma, id) != 0) {
```
One last question. Due to software or hardware bugs, this bit may not be able to set to the desired state. Using an endless while may cause hardware stuck here, especially when we are in the starting of the kernel. I think here we should wait 1~5 seconds, report the error and return. What do you think?

cybertale commented on drivers/dma/dma_stm32.c:
```diff
@@ -493,9 +495,11 @@ DMA_STM32_EXPORT_API int dma_stm32_reload(const struct device *dev, uint32_t id,
-	stm32_dma_disable_stream(dma, id);
 	stream = &config->streams[id];
 
+	while (stm32_dma_disable_stream(dma, id) != 0) {
```
Here, too.

lowlander replied:
Now used the available `dma_stm32_disable_stream` function, that has a finite loop, so now it should not hang forever if something goes wrong.

gmarull reviewed (approved):
LGTM. Tested on nucelo_f446re (dma loopback test).

cybertale reviewed (approved):
Thanks!

lowlander commented:
So everything rebased, and CI also happy again, now quickly merge it before someone breaks it again ;-)

cybertale commented:
@nashif @galak Please help merge this

erwango commented:
@lowlander since you're making significant changes to this dma driver, would you be interested to be added as codewoner along with @cybertale to be notified of further changes on this driver ?

lowlander commented:
> @lowlander since you're making significant changes to this dma driver, would you be interested to be added as codewoner along with @cybertale to be notified of further changes on this driver ?

@erwango feel free to add me, if @cybertale doesn't have a problem with it.

erwango commented:
> @erwango feel free to add me, if @cybertale doesn't have a problem with it.

The more, the merrier.
Point of being set as codeowner is that you'll be automatically notified and asked for reviews for any change on the targeted files. As it could happen that for any reason a codeowner could not have free cycles to spend on zephyr when the need arise, having more people involved is a plus (until a certain point).
